File: Lab2/apz-pzpi-22-9-aleksiienko-yevhenii-lab2/data/models/Device.kt
package com.smartmonitoringapplication.app.data.models

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
data class Device(
    val deviceId: String,
    val userId: User,
    val name: String,
    val description: String? = null,
    val location: String,
    val deviceType: String = "monitoring_station",
    val firmware: Firmware,
    val configuration: DeviceConfiguration,
    val status: DeviceStatus,
    val statistics: DeviceStatistics,
    val isActive: Boolean = true,
    val createdAt: String,
    val updatedAt: String
) : Parcelable

data class DeviceUser(
    val _id: String,
    val username: String,
    val email: String
)

@Parcelize
data class Firmware(
    val version: String = "1.0.0",
    val lastUpdate: String
) : Parcelable

@Parcelize
data class DeviceConfiguration(
    val sensors: SensorConfiguration,
    val alerts: AlertConfiguration
) : Parcelable

@Parcelize
data class SensorConfiguration(
    val temperature: TemperatureSensorConfig,
    val motion: MotionSensorConfig
) : Parcelable

@Parcelize
data class TemperatureSensorConfig(
    val enabled: Boolean = true,
    val interval: Int = 5000,
    val thresholds: TemperatureThresholds
) : Parcelable

@Parcelize
data class TemperatureThresholds(
    val warning: Double = 38.0,
    val critical: Double = 40.0
) : Parcelable

@Parcelize
data class MotionSensorConfig(
    val enabled: Boolean = true,
    val sensitivity: String = "medium",
    val zones: List<MotionZone> = emptyList()
) : Parcelable

@Parcelize
data class MotionZone(
    val id: Int,
    val name: String,
    val enabled: Boolean = true
) : Parcelable

@Parcelize
data class AlertConfiguration(
    val sound: Boolean = true,
    val frequency: Int = 7000
) : Parcelable

@Parcelize
data class DeviceStatus(
    val isOnline: Boolean = false,
    val lastSeen: String,
    val battery: BatteryStatus? = null,
    val wifi: WifiStatus,
    val memory: MemoryStatus? = null
) : Parcelable

@Parcelize
data class BatteryStatus(
    val level: Int? = null,
    val isCharging: Boolean = false
) : Parcelable

@Parcelize
data class WifiStatus(
    val connected: Boolean = false,
    val signalStrength: Int? = null,
    val ssid: String? = null
) : Parcelable

@Parcelize
data class MemoryStatus(
    val used: Long? = null,
    val free: Long? = null
) : Parcelable

@Parcelize
data class DeviceStatistics(
    val totalUptime: Long = 0,
    val totalAlerts: Int = 0,
    val lastReboot: String? = null,
    val dataPoints: DataPointsCount
) : Parcelable

@Parcelize
data class DataPointsCount(
    val today: Int = 0,
    val thisWeek: Int = 0,
    val thisMonth: Int = 0
) : Parcelable

// API Response models
data class DevicesResponse(
    val success: Boolean,
    val message: String? = null,
    val data: DevicesData? = null
)

data class DevicesData(
    val devices: List<Device>,
    val pagination: Pagination
)

data class DeviceResponse(
    val success: Boolean,
    val message: String? = null,
    val data: DeviceData? = null
)

data class DeviceData(
    val device: Device,
    val recentData: List<SensorData>? = null
)

data class Pagination(
    val current: Int,
    val pages: Int,
    val total: Int,
    val limit: Int
)

// Request models
data class CreateDeviceRequest(
    val name: String,
    val description: String? = null,
    val location: String,
    val deviceType: String = "monitoring_station",
    val configuration: DeviceConfiguration? = null
)

data class UpdateDeviceRequest(
    val name: String? = null,
    val description: String? = null,
    val location: String? = null,
    val configuration: DeviceConfiguration? = null,
    val isActive: Boolean? = null
)

data class CommandRequest(
    val command: String,
    val parameters: Map<String, Any> = emptyMap()
)

File: Lab4/apz-pzpi-22-9-aleksiienko-yevhenii-lab4/models/Device.js
const mongoose = require('mongoose');

const deviceSchema = new mongoose.Schema({
  deviceId: {
    type: String,
    required: true,
    unique: true,
    // Format: YYYYMMDD-userID-5randmSymbols
    match: /^\d{8}-[a-zA-Z0-9]+-[a-zA-Z0-9]{5}$/,
    trim: true
  },
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  name: {
    type: String,
    required: true,
    trim: true,
    maxlength: 100
  },
  description: {
    type: String,
    trim: true,
    maxlength: 500
  },
  location: {
    type: String,
    required: true,
    trim: true,
    maxlength: 100
  },
  deviceType: {
    type: String,
    enum: ['monitoring_station', 'sensor_hub', 'smart_home'],
    default: 'monitoring_station'
  },
  firmware: {
    version: {
      type: String,
      default: '1.0.0'
    },
    lastUpdate: {
      type: Date,
      default: Date.now
    }
  },
  configuration: {
    sensors: {
      temperature: {
        enabled: {
          type: Boolean,
          default: true
        },
        interval: {
          type: Number,
          default: 5000, // milliseconds
          min: 1000,
          max: 300000
        },
        thresholds: {
          warning: {
            type: Number,
            default: 38.0
          },
          critical: {
            type: Number,
            default: 40.0
          }
        }
      },
      motion: {
        enabled: {
          type: Boolean,
          default: true
        },
        sensitivity: {
          type: String,
          enum: ['low', 'medium', 'high'],
          default: 'medium'
        },
        zones: [{
          id: {
            type: Number,
            required: true
          },
          name: {
            type: String,
            required: true
          },
          enabled: {
            type: Boolean,
            default: true
          }
        }]
      }
    },
    alerts: {
      sound: {
        type: Boolean,
        default: true
      },
      frequency: {
        type: Number,
        default: 7000, // Hz
        min: 1000,
        max: 10000
      }
    }
  },
  status: {
    isOnline: {
      type: Boolean,
      default: false
    },
    lastSeen: {
      type: Date,
      default: Date.now
    },
    battery: {
      level: {
        type: Number,
        min: 0,
        max: 100
      },
      isCharging: {
        type: Boolean,
        default: false
      }
    },
    wifi: {
      connected: {
        type: Boolean,
        default: false
      },
      signalStrength: {
        type: Number,
        min: -100,
        max: 0
      },
      ssid: {
        type: String,
        trim: true
      }
    },
    memory: {
      used: {
        type: Number
      },
      free: {
        type: Number
      }
    }
  },
  statistics: {
    totalUptime: {
      type: Number,
      default: 0 // seconds
    },
    totalAlerts: {
      type: Number,
      default: 0
    },
    lastReboot: {
      type: Date
    },
    dataPoints: {
      today: {
        type: Number,
        default: 0
      },
      thisWeek: {
        type: Number,
        default: 0
      },
      thisMonth: {
        type: Number,
        default: 0
      }
    }
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true
});

// Indexes (убираем дублированный индекс deviceId так как он уже unique: true)
deviceSchema.index({ userId: 1 });
deviceSchema.index({ 'status.isOnline': 1 });
deviceSchema.index({ isActive: 1 });
deviceSchema.index({ createdAt: -1 });

// Virtual for getting device age
deviceSchema.virtual('age').get(function() {
  return Math.floor((Date.now() - this.createdAt) / (1000 * 60 * 60 * 24)); // days
});

// Method to update last seen
deviceSchema.methods.updateLastSeen = function() {
  this.status.lastSeen = new Date();
  this.status.isOnline = true;
  return this.save();
};

// Method to mark device as offline
deviceSchema.methods.markOffline = function() {
  this.status.isOnline = false;
  return this.save();
};

// Method to increment alert count
deviceSchema.methods.incrementAlerts = function() {
  this.statistics.totalAlerts += 1;
  return this.save();
};

// Static method to find online devices
deviceSchema.statics.findOnline = function() {
  return this.find({ 'status.isOnline': true, isActive: true });
};

// Static method to find devices by user
deviceSchema.statics.findByUser = function(userId) {
  return this.find({ userId, isActive: true });
};

// Static method to find devices that haven't been seen recently
deviceSchema.statics.findStale = function(minutes = 10) {
  const staleTime = new Date(Date.now() - minutes * 60 * 1000);
  return this.find({
    'status.lastSeen': { $lt: staleTime },
    'status.isOnline': true,
    isActive: true
  });
};

module.exports = mongoose.model('Device', deviceSchema);